heka/plugins/s3splitfile/s3splitfile_common.go

```diff
@@ -214,16 +221,224 @@ func LoadSchema(schemaFileName string) (schema Schema, err error) {
+			// Response is not truncated, so we're done.
+			done = true
+		}
+
+		if level >= len(schema.Fields) {
+			// We are past all the dimensions - encountered items are now
+			// S3 key names. We ignore any further prefixes and assume that the
+			// specified schema is correct/complete.
+			for _, k := range response.Contents {
+				marker = k.Key
+				kc <- S3ListResult{k, nil}
+			}
+		} else {
+			// We are still looking at prefixes. Recursively list each one that
+			// matches the specified schema's allowed values.
+			for _, pf := range response.CommonPrefixes {
+				// Get just the last piece of the prefix to check it as a
+				// dimension. If we have '/foo/bar/baz', we just want 'baz'.
+				stripped := pf[len(prefix) : len(pf)-1]
+				allowed := schema.Dims[schema.Fields[level]].IsAllowed(stripped)
+				marker = pf
+				if allowed {
+					FilterS3(bucket, pf, level+1, schema, kc)
+				}
+			}
+		}
+	}
+
+	if level == 0 {
+		// We traverse the tree in depth-first order, so once we've reached the
+		// end at the root (level 0), we know we're done.
+		// Note that things could be made faster by parallelizing the recursive
+		// listing, but we would need some other mechanism to know when to close
+		// the channel?
+		close(kc)
+	}
+	return
+}
+
+// Encapsulates a single record within an S3 file, allowing detection of errors
+// along the way.
+type S3Record struct {
+	BytesRead int
+	Record    []byte
+	Err       error
+}
+
+// List the contents of the given bucket, sending matching filenames to a
+// channel which can be read by the caller.
+func S3FileIterator(bucket *s3.Bucket, s3Key string) <-chan S3Record {
+	recordChannel := make(chan S3Record, fileBatchSize)
+	go ReadS3File(bucket, s3Key, recordChannel)
+	return recordChannel
+}
+
+func makeS3Record(bytesRead int, data []byte, err error) (result S3Record) {
+	r := S3Record{}
+	r.BytesRead = bytesRead
+	r.Err = err
+	r.Record = make([]byte, len(data))
+	copy(r.Record, data)
+	return r
+}
+
+// TODO: duplicated from heka-cat
+func makeSplitterRunner() (SplitterRunner, error) {
+	splitter := &HekaFramingSplitter{}
+	config := splitter.ConfigStruct()
+	err := splitter.Init(config)
+	if err != nil {
+		return nil, fmt.Errorf("Error initializing HekaFramingSplitter: %s", err)
+	}
+	srConfig := CommonSplitterConfig{}
+	sRunner := NewSplitterRunner("HekaFramingSplitter", splitter, srConfig)
+	return sRunner, nil
+}
+
+func ReadS3File(bucket *s3.Bucket, s3Key string, recordChan chan S3Record) {
+	defer close(recordChan)
+	sRunner, err := makeSplitterRunner()
+	if err != nil {
+		recordChan <- S3Record{0, []byte{}, err}
+		return
+	}
+	reader, err := bucket.GetReader(s3Key)
+	if err != nil {
+		recordChan <- S3Record{0, []byte{}, err}
+		return
+	}
+	defer reader.Close()
+
+	var size int64
+
+	done := false
+	for !done {
+		n, record, err := sRunner.GetRecordFromStream(reader)
+		size += int64(n)
+
+		if err != nil {
+			if err == io.EOF {
+				if len(record) != 0 {
```

Comment:
I wanted to make sure that the Printf call below happened if something weird was going on with EOF and GetRemainingData().  If record *is* empty, then we don't care what's in GetRemainingData(), since it's an incomplete message.  If it is not empty, we should use it, and *then* check GetRemainingData (and not overwrite record at this point).  Will fix.